### PR TITLE
🐛 Fix int or string pattern application

### PR DESCRIPTION
**Pull request by JoelSpeed**
This is a follow up to fix #608 which I hadn't appropriately tested before proposing it 😞 

I tried this out this morning and worked out that it wasn't working, there are two reasons for this.

The condition check was wrong, we meant to check for the schema *not* being an IntOrString.

The second issue is that the additional information for the type that the generator applies (eg setting XIntOrString=true) is not done until way later in the generation than the markers being applied. Looking at the existing code, the "fix" for this seems to be that you have to apply an additional marker, eg the Type marker, that sets the required fields early.

So this introduces a new marker `// +kubebuilder:validation:XIntOrString` which marks the type as an IntOrString early to allow the pattern marker to pick this up. Not sure if this is the best approach given Solly's old notes, but it seems to be in-keeping with the rest of the markers

/CC @alvaroaleman @vincepri

### CODE REVIEW

**Comment by alvaroaleman** [comment]
@JoelSpeed can we properly test this by extending pkg/crd/testdata/cronjob_types.go with a field of type IntOrString and a pattern?

**Comment by JoelSpeed** [comment]
> @JoelSpeed can we properly test this by extending pkg/crd/testdata/cronjob_types.go with a field of type IntOrString and a pattern?

Yes of course! I did have a quick scan for something testing the CRD generation but didn't find this, will take a look at adding a test today

File: pkg/crd/testdata/testdata.kubebuilder.io_cronjobs.yaml [hunk 1]
```diff
@@ -125,6 +125,16 @@ spec:
...
+              intOrStringWithAPattern:
+                anyOf:
+                - type: integer
+                - type: string
+                description: This tests that an IntOrString can also have a pattern
+                  attached to it. This can be useful if you want to limit the string
+                  to a perecentage or integer. The XIntOrString marker is a requirement
+                  for having a pattern on this type.
+                pattern: ^((100|[0-9]{1,2})%|[0-9]+)$
+                x-kubernetes-int-or-string: true
```
Comment (alvaroaleman) [inline pkg/crd/testdata/testdata.kubebuilder.io_cronjobs.yaml:137]: Did you btw if the apiserver correctly deals with this (i.E. applies the pattern when you submit a string but not when you submit an int)?

Comment (JoelSpeed) [reply pkg/crd/testdata/testdata.kubebuilder.io_cronjobs.yaml:137]: Yeah I tested this manually in kind. I could apply any number I wanted but was restricted if I added a percent sign

**Review by alvaroaleman — APPROVED** [review]
/label tide/merge-method-squash
/lgtm
/assign @vincepri

**Comment by JoelSpeed** [comment]
Had to fix a typo in the comment on the new marker, no other changes

**Pull request merged**